---
engine: knitr
---

# GWAS

## Setup

```{bash}
[ -f data/gwas.imputed.chr20.dose.vcf.gz ] || wget -q -P data https://genepi.i-med.ac.at/downloads/imputation/gwas.imputed.chr20.dose.vcf.gz
mkdir -p output
```

- Install PLINK2

## nf-gwas


## PLINK2

### Without PCA

```{bash}
plink2 --vcf data/gwas.imputed.chr20.dose.vcf.gz \
  --pheno data/phenotypes.txt \
  --pheno-name pheno_1 \
  --logistic allow-no-covars \
  --out output/gwas
```

```{r}
library(qqman)

pheno_1 <- read.table("output/gwas.pheno_1.glm.logistic.hybrid", header=TRUE, comment.char="")
pheno_1 <- pheno_1[!is.na(pheno_1$P), ]

manhattan(x = pheno_1, chr = "X.CHROM", bp = "POS", p = "P", snp="ID")
qq(pheno_1$P)
```

### PCA

```{r}
library(ggplot2)

pca_data <- read.table("data/covariates.txt", header=TRUE, comment.char="")
#phenotypes <- read.table("data/phenotypes.txt", header=TRUE, comment.char="")

# merge pca data with phenotypes by column IID
#phenotypes <- merge(pca_data, phenotypes, by = "IID")

#ggplot(phenotypes, aes(x = PC1, y = PC2, color = super_pop)) + geom_point()
ggplot(pca_data, aes(x = PC1, y = PC2)) + geom_point() + theme_minimal()

```

### With PCA

```{bash}
plink2 --vcf data/gwas.imputed.chr20.dose.vcf.gz \
  --pheno data/phenotypes.txt \
  --pheno-name pheno_1 \
  --covar data/covariates.txt \
  --logistic hide-covar \
  --out output/gwas.covar
```

```{r}
pheno_1 <- read.table("output/gwas.covar.pheno_1.glm.logistic.hybrid", header=TRUE, comment.char="")
pheno_1 <- pheno_1[!is.na(pheno_1$P), ]
manhattan(x = pheno_1, chr = "X.CHROM", bp = "POS", p = "P", snp="ID")
qq(pheno_1$P)
```